skfem/mesh/mesh.py

```diff
@@ -147,50 +152,45 @@ def refine(self, arg: Optional[Union[int, ndarray]] = None):
             self._adaptive_refine(arg)
         else:
             raise NotImplementedError("The parameter type not supported.")
+        return self
```

Comment:
Hmm, following the discussion in #536, I wonder whether having `Mesh.refine` 
* mutate `self` *and* 
* return the mutated self 

could cause just the kind of confusion that I wish to avoid with mutation in general; i.e. if the only usage of `Mesh.refine` is
```python
m.refine()
```
then it's pretty clear that it's mutating but if I see
```python
m_new = m_old.refine()
```
then I might forget that it's impure and forget about the mutation and get surprised if I re-access `m_old` and find it changed.  (Maybe in some kind of manual multigrid?)

Could this be why the Python Standard Library offers the pure function [`sorted()`](https://docs.python.org/3/library/functions.html#sorted) and the in-place method [`list.sort()`](https://docs.python.org/3/library/stdtypes.html#list.sort)?  Should the in-place method `Mesh.refine` be retained (for full backward-compatibility) and a pure method `Mesh.refined` be introduced?   (If so, in due course, the examples, docs, tests, &c., might migrate to the latter, with the former perhaps eventually being scheduled for deprecation.)

(In the case of `refine`, [I don't think](https://github.com/kinnala/scikit-fem/issues/536#issuecomment-758351972) it gains any efficiency by working in place so it might well be redefined in terms of the proposed pure `Mesh.refined`.)

Reply:
I kind of like this naming: `refine` vs. `refined`, `scale` vs. `scaled`, `translate` vs. `translated`. I'll look into it after I stop being busy with a peer review appointment.

Reply:
For now I created `refined` basically as
```python
mesh = self.copy()
mesh.refine()
return mesh
```
because we still aim to be backwards-compatible.
Later when we are ready to drop `Mesh.refine` we can refactor it differently.